torch/distributed/pipelining/_backward.py

```diff
@@ -223,7 +207,7 @@ def hook(grad_inputs):
 
 
 def stage_backward_weight(
-    weights: Iterator[Parameter], param_groups: List[Dict[str, Any]]
+    weights: Iterator[Parameter], param_groups: List[Dict[str, Any]], retain_graph=False
```

Comment:
It the `retain_graph` flag is just for test purpose, shall we make it `_retain_graph` and add a banner saying "Test only; don't use"?

Reply:
I added `retain_graph` to better align with the `.backward()` API in the case that someone wants to perform multiple backwards (double backward) and accumulate the gradients. I was using this in testing and I dont think anyone besides us is using the `stage_backward_input` and `stage_backward_weight`, but they dont necessarily need to only apply to stages and could be used as a more general API.